packages/playwright-test/src/plugins/webServerPlugin.ts

```diff
@@ -13,66 +13,102 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 import http from 'http';
 import https from 'https';
+import path from 'path';
 import net from 'net';
+
 import { debug } from 'playwright-core/lib/utilsBundle';
 import { raceAgainstTimeout } from 'playwright-core/lib/utils/timeoutRunner';
-import type { FullConfig } from './types';
 import { launchProcess } from 'playwright-core/lib/utils/processLauncher';
-import type { Reporter } from '../types/testReporter';
+
+import type { PlaywrightTestConfig, TestPlugin } from '../types';
+import type { Reporter } from '../../types/testReporter';
+
+export type WebServerPluginOptions = {
+  command: string;
+  url: string;
+  ignoreHTTPSErrors?: boolean;
+  timeout?: number;
+  reuseExistingServer?: boolean;
+  cwd?: string;
+  env?: { [key: string]: string; };
+};
+
+interface InternalWebServerConfigOptions extends Omit<WebServerPluginOptions, 'url'> {
+  setBaseURL: boolean;
+  url?: string;
+  port?: number;
+}
 
 const DEFAULT_ENVIRONMENT_VARIABLES = {
   'BROWSER': 'none', // Disable that create-react-app will open the page in the browser
 };
 
 const debugWebServer = debug('pw:webserver');
 
-type WebServerConfig = NonNullable<FullConfig['webServer']>;
-
-export class WebServer {
+export class InternalWebServerPlugin implements TestPlugin {
   private _isAvailable: () => Promise<boolean>;
   private _killProcess?: () => Promise<void>;
   private _processExitedPromise!: Promise<any>;
+  private _config: InternalWebServerConfigOptions;
+  private _reporter: Reporter;
+
+  constructor(config: InternalWebServerConfigOptions, reporter: Reporter) {
+    this._reporter = reporter;
+    this._config = { ...config };
+    this._config.setBaseURL = this._config.setBaseURL ?? true;
+    this._isAvailable = getIsAvailableFunction(config, this._reporter.onStdErr?.bind(this._reporter));
+  }
+
+  get name() {
+    const target = this._config.url || `http://localhost:${this._config.port}`;
```

Comment:
For example, legacy plugin could wrap this new shiny plugin and convert port into url in its own layer.

Reply:
Agree—once all these settle, I plan to make it even better including your suggestion. We always assert that we get an OK response back when using a URL, whereas port option (of legacy) just checks something is listening there, so they are not strictly convertable.

However, in new shiny version, I want to have `readinessCheck: Promise<bool>` which allows user to provide custom logic which could be anything from:

1. check a URL for a 200.
2. check a file is on disk.
3. check a DB is ready and migrated.

or, for legacy behavior, it's just a function that `isPortListening`.